backend/multimodalexplorer/functions/fetch_embed.py

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+import logging
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+import pandas as pd
+
+from multimodalexplorer.utils.helpers import get_file_path
+from multimodalexplorer.utils.utils import parse_arguments, select_params
+
+# Set up logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def fetch_embed() -> Optional[List[Dict[str, Any]]]:
+    """
+    Load UMAP embeddings from the stored file and merge them with raw data.
+
+    Returns:
+        Optional[List[Dict[str, Any]]]: The merged data containing UMAP embeddings or None if an error occurred.
+    """
+    # Parse command line arguments
+    args = parse_arguments()
+    # Select relevant parameters
+    params = select_params(args, ["umap_file", "raw_data_file"])
+    umap_file, raw_data_file = params
+
+    # Get file paths for UMAP embeddings and raw data
+    dir_path, ext = umap_file.values()
+    file_path_embed = get_file_path(dir_path, ext)
+
+    dir_path, ext = raw_data_file.values()
+    file_path_raw = get_file_path(dir_path, ext)
+
+    try:
+        # Load raw data
+        raw_data = pd.read_csv(file_path_raw, sep="\t")
+
+        # Load UMAP embeddings
+        with open(file_path_embed, "rb") as file:
+            embeddings = np.load(file)
+
+        # Merge raw data with UMAP embeddings
+        merged_data = pd.concat([raw_data, pd.DataFrame(embeddings)], axis=1)
+
+        # Rename columns
+        merged_data.rename(
+            columns={
+                merged_data.columns[-3]: "x",
+                merged_data.columns[-2]: "y",
+                merged_data.columns[-1]: "cluster",
+            },
+            inplace=True,
+        )
+
+        # Convert merged data to list of dictionary records
+        merged_embed_data = merged_data.to_dict("records")
```

Comment:
Here, I'm currently merging the 'umap_embedding' NumPy array with the raw data in CSV format within a dictionary. This combined data is what I'm sending to the client to fetch the embeddings. The purpose behind this approach is to gather all the embedding data upfront and have it ready to draw on the canvas when the visualizer is zoomed in on the user interface (UI). This helps avoid making repeated calls to the UI.

Initially, I tried making another call to the UI when the visualizer is sufficiently zoomed, but this resulted in multiple calls being rendered as the user interacts with the window. I'm unsure if the current implementation is the most efficient way to handle this.

Reply:
I am not sure I understand. What do you call raw data? Is this the original text before embedding? Where is this raw_data_file created.

Reply:
Yes, this is the initial text prior to embedding. The logic for saving raw_data is contained within the _save_data function in the process_dataset script.
This file is generated and the raw_data is appended to it concurrently while I encode and save the embeddings in batches and chunks. It's formatted as a TSV file with two columns: data and media_type.